File: backend/lab-resource-platform/src/main/java/com/lab/resource/service/BookingService.java
package com.lab.resource.service;

import com.lab.resource.entity.Booking;

import java.time.LocalDate;
import java.util.List;

public interface BookingService {

    // =====================================================
    // BOOKING CRUD
    // =====================================================

    Booking createBooking(Booking booking);

    List<Booking> getAllBookings();

    Booking getBookingById(Long id);

    Booking updateBooking(Long id, Booking booking);

    void deleteBooking(Long id);

    // =====================================================
    // BOOKING WORKFLOW
    // =====================================================

    Booking returnEquipment(Long bookingId);

    List<Booking> getWaitingQueue(Long equipmentId);

    Booking approveBooking(Long bookingId);

    Booking rejectBooking(Long bookingId);

    // =====================================================
    // DASHBOARD STATISTICS
    // =====================================================

    long getTotalBookings();

    long getBookedCount();

    long getReturnedCount();

    long getWaitingCount();

    long getActiveBookings();

    long getTodayBookings();

    long getTodayReturns();

    // =====================================================
    // REPORTS
    // =====================================================

    List<Booking> getBookingsByStatus(String status);

    List<Booking> getActiveBookingList();

    List<Booking> getAllBookingsNewestFirst();

    List<Booking> getBookingsBetweenDates(
            LocalDate startDate,
            LocalDate endDate
    );

    List<Booking> getReturnsBetweenDates(
            LocalDate startDate,
            LocalDate endDate
    );

}

File: frontend/src/services/bookingService.js
import api from "./api";

// ===================================
// Get All Bookings
// ===================================

export const getAllBookings = async () => {
    const response = await api.get("/bookings");
    return response.data;
};

// ===================================
// Get Booking By ID
// ===================================

export const getBookingById = async (id) => {
    const response = await api.get(`/bookings/${id}`);
    return response.data;
};

// ===================================
// Create Booking
// ===================================

export const createBooking = async (booking) => {
    const response = await api.post("/bookings", booking);
    return response.data;
};

// ===================================
// Update Booking
// ===================================

export const updateBooking = async (id, booking) => {
    const response = await api.put(`/bookings/${id}`, booking);
    return response.data;
};

// ===================================
// Delete Booking
// ===================================

export const deleteBooking = async (id) => {
    const response = await api.delete(`/bookings/${id}`);
    return response.data;
};

// ===================================
// Return Equipment
// ===================================

export const returnEquipment = async (id) => {
    const response = await api.put(`/bookings/${id}/return`);
    return response.data;
};

// ===================================
// Waiting Queue
// ===================================

export const getWaitingQueue = async (equipmentId) => {
    const response = await api.get(`/bookings/queue/${equipmentId}`);
    return response.data;
};

// ===================================
// Approve Booking
// ===================================

export const approveBooking = async (id) => {
    const response = await api.put(`/bookings/${id}/approve`);
    return response.data;
};

// ===================================
// Reject Booking
// ===================================

export const rejectBooking = async (id) => {
    const response = await api.put(`/bookings/${id}/reject`);
    return response.data;
};

// ===================================
// Dashboard Statistics
// ===================================

export const getTotalBookings = async () => {
    const response = await api.get("/bookings/count");
    return response.data;
};

export const getBookedCount = async () => {
    const response = await api.get("/bookings/count/booked");
    return response.data;
};

export const getReturnedCount = async () => {
    const response = await api.get("/bookings/count/returned");
    return response.data;
};

export const getWaitingCount = async () => {
    const response = await api.get("/bookings/count/waiting");
    return response.data;
};

export const getActiveBookingsCount = async () => {
    const response = await api.get("/bookings/count/active");
    return response.data;
};

// ===================================
// Reports
// ===================================

export const getBookingsByStatus = async (status) => {
    const response = await api.get(`/bookings/status/${status}`);
    return response.data;
};

export const getActiveBookings = async () => {
    const response = await api.get("/bookings/active");
    return response.data;
};

export const getLatestBookings = async () => {
    const response = await api.get("/bookings/latest");
    return response.data;
};

export const getBookingsBetweenDates = async (startDate, endDate) => {
    const response = await api.get("/bookings/between", {
        params: {
            startDate,
            endDate,
        },
    });
    return response.data;
};

export const getReturnsBetweenDates = async (startDate, endDate) => {
    const response = await api.get("/bookings/returns", {
        params: {
            startDate,
            endDate,
        },
    });
    return response.data;
};